File: maximum-and-minimum-el/C/maximum_and_minimal_el.c
// Find the mininum and maximum element in a array

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <limits.h>

#define ARR_SIZE(arr) (sizeof(arr) / sizeof(arr[0]))

bool find_min(int *arr, size_t arr_size, int *result)
{
    if (arr == NULL || arr_size == 0) {
        fprintf(stderr, "Invalid array\n");
        return false;
    }
    int min = INT_MAX;
    for (size_t i = 0; i < arr_size; i++) {
        if (arr[i] < min) 
            min = arr[i];
    }
    *result = min;
    return true;
}

bool find_max(int *arr, size_t arr_size, int *result)
{
    if (arr == NULL || arr_size == 0) {
        fprintf(stderr, "Invalid array\n");
        return false;
    }
    int max = INT_MIN;
    for (size_t i = 0; i < arr_size; i++) {
        if (arr[i] > max) 
            max = arr[i];
    }
    *result = max;
    return true;
}
int main(void)
{
    // Array for testing: 
    int arr[] = { 10, 20, -5, -10, 50 };
    size_t arr_size = ARR_SIZE(arr);
    int result;

    if (find_min(arr, arr_size, &result)) 
        printf("Mininum value: %d\n", result);
    
    if (find_max(arr, arr_size, &result)) 
        printf("Maximum value: %d\n", result);
    return EXIT_SUCCESS;
}

File: maximum-and-minimum-el/JS/maximumAndMinimalEl.js
// Find the mininum and maximum element in a array

// --
// Arrays for testing
// --

const arr = [10, 20, -5, -10, 50];
const arr2 = [10, 20, -5, -10, 50];

// This is a linear search and time complexity for it is O(n);
// where n is the total number of elements in the original array
// We could use a sorting builtin method and return the first element but that would not be optimal
// since the best general-purpse sorting algorithm time complexity is O(n log n)
function findMin(arr) {
    // General checking     
    // this is saying: array needs to be an actual array 
    // it needs to have at least one element  
    // every element need to be a number
    if (!Array.isArray(arr) || arr.length === 0 || !arr.every(e => typeof(e) === 'number')) {
        console.error('Invalid array');
        return undefined;
    }
    let min = +Infinity;
    
    for (const el of arr) {
        if (el < min) min = el;
    }
    return min;
}

let result = findMin(arr);
if (result !== undefined) console.log('The mininum element is:', result); // -> -10

function findMax(arr) {
    if (!Array.isArray(arr) || arr.length === 0 || !arr.every(e => typeof(e) === 'number')) {
        console.error('Invalid array');
        return undefined;
    }
    let max = -Infinity;
    
    for (const el of arr) {
        if (el > max) max = el;
    }
    return max;
}

result = findMax(arr);
if (result !== undefined) console.log('The maximum element is:', result); // -> 50
